Fix serialization of UPath

normanrz opened the pull request:
Currently, pickled instances of `UPath` will be un-pickled as `pathlib.PosixPath`. This PR fixes that by implementing `__reduce__` and `__setstate__` methods.

Fixes #44

andrewfulton9 commented on upath/tests/test_core.py:
```diff
@@ -137,3 +137,24 @@ def test_create_from_type(path, storage_options, module, object_type):
+
+
+def test_pickling():
+    path = UPath("s3://bucket/folder", storage_options={"anon": True})
```
S3 tests are currently skipped in the CI/CD for this so `s3fs` isn't installed. If you change these paths to 'gcs://' they should pass.

andrewfulton9 commented:
Thanks for the PR, @normanrz! I made a comment to get the CI/CD tests passing. If you make that change, I get this merged ASAP.

normanrz commented:
I fixed the tests. I also refactored the `/` (i.e. `__truediv__`) operator so that the following two expressions produce the same representation. Previously, there were differences in `_parts` and `_url.path`, which are cloaked in `_format_parsed_parts`.
```python
path_a = UPath("gcs://bucket/folder")
path_b = UPath("gcs://bucket") / "folder"
```

Also, the `mock:` tests gave me a bit of trouble on Windows, because the file path that is used is Windows-style path, but `UPath` implements `_posix_flavour`, which expects a `/` at the beginning in some methods.

andrewfulton9 commented:
> I also refactored the `/` (i.e. `__truediv__`) operator so that the following two expressions produce the same representation. Previously, there were differences in `_parts` and `_url.path`, which are cloaked in `_format_parsed_parts`.
> 
> ```python
> path_a = UPath("gcs://bucket/folder")
> path_b = UPath("gcs://bucket") / "folder"
> ```

Great! Thank you!

> Also, the `mock:` tests gave me a bit of trouble on Windows, because the file path that is used is Windows-style path, but `UPath` implements `_posix_flavour`, which expects a `/` at the beginning in some methods.

Interesting. I'm surprised that hasn't shown up on the CI/CD windows tests. Thanks for adding the fix though. I'll go ahead and merge this.

normanrz commented:
Thanks for merging! Could you please also create a new pypi release?

normanrz commented:
Thanks for releasing!
